### PR TITLE
Make parallel restore configurable

### PR DESCRIPTION
**Pull request by qiuming-best**
Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)
https://github.com/vmware-tanzu/velero/issues/7479
# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: pkg/builder/restore_builder.go [hunk 1]
```diff
@@ -171,9 +171,3 @@ func (b *RestoreBuilder) ItemOperationTimeout(timeout time.Duration) *RestoreBui
 	b.object.Spec.ItemOperationTimeout.Duration = timeout
 	return b
 }
-
-// WriteSparseFiles sets the Restore's uploader write sparse files
-func (b *RestoreBuilder) WriteSparseFiles(val bool) *RestoreBuilder {
```
Comment (Lyndon-Li) [inline pkg/builder/restore_builder.go:176]: Why do we remove this helper func?

File: pkg/uploader/kopia/snapshot.go [hunk 2]
```diff
@@ -419,9 +421,17 @@ func Restore(ctx context.Context, rep repo.RepositoryWriter, progress *Progress,
 		if writeSparseFiles {
 			fsOutput.WriteSparseFiles = true
 		}
+
+		concurrency, err := uploaderutil.GetRestoreConcurrency(uploaderCfg)
+		if err != nil {
+			return 0, 0, errors.Wrap(err, "failed to get parallel restore uploader config")
+		}
+		if restoreConcurrency > 0 {
```
Comment (Lyndon-Li) [inline pkg/uploader/kopia/snapshot.go:429]: `if restoreConcurrency > 0`  -> `if concurrency > 0`

File: design/Implemented/velero-uploader-configuration.md [hunk 3]
```diff
@@ -177,5 +177,54 @@ Roughly, the process is as follows:
...
+```
+
+### UploaderConfig
+below the sub-option parallel is added into UploaderConfig:
+
+```go
+	type UploaderConfigForRestore struct {
+		// ParallelFilesDownload is the number of parallel for restore.
+		// +optional
+		ParallelFilesDownload int `json:"parallel,omitempty"`
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/Implemented/velero-uploader-configuration.md]: Updated

File: design/Implemented/velero-uploader-configuration.md [hunk 4]
```diff
@@ -177,5 +177,54 @@ Roughly, the process is as follows:
 4. Each respective controller within the CRs calls the uploader, and the WriteSparseFiles from map in CRs is passed to the uploader.
 5. When the uploader subsequently calls the Kopia API, it can use the WriteSparseFiles to set the WriteSparseFiles parameter, and if the uploader calls the Restic command it would append `--sparse` flag within the restore command.
 
+### Parallel Restore
+Setting the parallelism of restore operations can improve the efficiency and speed of the restore process, especially when dealing with large amounts of data.
+
+### Velero CLI
+The Velero CLI will support a --parallel-files-download flag, allowing users to set the parallelism value when creating restores. when no value specified, the value of it would be the number CPU cores allocated to Velero node agent pod.
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/Implemented/velero-uploader-configuration.md]: Updated

File: pkg/cmd/cli/restore/create.go [hunk 5]
```diff
@@ -151,6 +152,8 @@ func (o *CreateOptions) BindFlags(flags *pflag.FlagSet) {
 
 	f = flags.VarPF(&o.WriteSparseFiles, "write-sparse-files", "", "Whether to write sparse files during restoring volumes")
 	f.NoOptDefVal = cmd.TRUE
+
+	flags.IntVar(&o.ParallelFilesDownload, "parallel-files-download", 0, "The number of restore operations to run in parallel. If set to 0, the default parallelism will be the number of CPUs allocated for node agent pod.")
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply pkg/cmd/cli/restore/create.go]: Updated

File: pkg/builder/restore_builder.go (hunk 1 above)
Comment (qiuming-best) [reply pkg/builder/restore_builder.go:176]: It's redundant and not used anywhere

File: pkg/uploader/kopia/snapshot.go [hunk 6]
```diff
@@ -436,7 +446,7 @@ func Restore(ctx context.Context, rep repo.RepositoryWriter, progress *Progress,
 	}
 
 	stat, err := restoreEntryFunc(kopiaCtx, rep, output, rootEntry, restore.Options{
-		Parallel:               runtime.NumCPU(),
```
Comment (blackpiglet) [inline pkg/uploader/kopia/snapshot.go:439]: A quick question.
Why the uploader parallel setting is not the same for backup and restore?

Comment (qiuming-best) [reply pkg/uploader/kopia/snapshot.go:439]: Do you mean that one setting for backup and restore?

Comment (blackpiglet) [reply pkg/uploader/kopia/snapshot.go:439]: Yes. The backup's uploader parallel setting defaults to 1, and the restore's uploader parallel setting defaults to the CPU number.

**Review by blackpiglet — APPROVED** [review]

**Review by Lyndon-Li — APPROVED** [review]

**Pull request merged**